### PR TITLE
Ensure coupon code times_used decrements on cancel

### LINKED ISSUE #167: Canceled order - reuse coupon by the same customer
Hi.
Magento problem for ages. 
User is making order, using coupon - one time. Order is canceled due to any problems with payment. Very often.
User tries to reorder - and try to use the coupon again. No luck

Can somebody fix this issue (feature) in Magento that is really annoying. Takes time for custome support, they are asking stuff to help, stuff must regenerate coupons and send them manually.
The goal to allow reuse coupon that was used for canceled order. Nothing more. Key is canceled order state.
Any options to fix this?
Related Stack Exchange topic
http://magento.stackexchange.com/questions/125849/coupon-code-should-apply-once-we-cancel-the-order
Solution to check:
http://magebase.com/magento-tutorials/quick-fix-coupon-codes-used-up-on-incomplete-transactions OR [removed bad link]

Thanks
Tomek

### PR DESCRIPTION
**Pull request by mattdavenport**
To implement this cleanly in the observer (without instantiating new connection objects, fetching records, etc.) I added a parameter `$decrement` to `updateCustomerCouponTimesUsed` which should be a non-breaking change.

fixes #167

### CODE REVIEW

**Comment by sreichel** [comment]
LGTM ... but can you please add some steps to reproduce?

**Comment by mattdavenport** [comment]
Sure thing!

1. In `Shopping Cart Price Rules`, add a new rule with a "Specific Coupon" code with > 0 uses per coupon.
2. In `Actions` tab, set a discount amount.
3. Add a product to cart. Apply promo code in cart. Checkout.
4. Upon checkout, it can be observed that `times_used` has been incremented in tables:
    - `salesrule_coupon`
    - `salesrule_coupon_usage`
    - `salesrule_customer`
5. In admin panel, cancel the order.
6. It can be observed that `times_used` has decremented to allow the coupon to be used again.

**Comment by colinmollenhour** [comment]
@tomekjordan @seansan may be interested in reviewing

File: app/code/core/Mage/SalesRule/Model/Observer.php [hunk 1]
```diff
@@ -132,6 +132,40 @@ public function sales_order_afterPlace($observer)
...
+            if ($code = $order->getCouponCode()) {
+                // Decrement coupon times_used
+                $coupon = Mage::getModel('salesrule/coupon')->loadByCode($code);
+                $coupon->setTimesUsed($coupon->getTimesUsed() - 1);
+                $coupon->save();
+
+
+                if ($customerId = $order->getCustomerId()) {
+                    // Decrement coupon_usage times_used
+                    Mage::getResourceModel('salesrule/coupon_usage')->updateCustomerCouponTimesUsed($customerId, $coupon->getId(), TRUE);
```
Comment (colinmollenhour) [inline app/code/core/Mage/SalesRule/Model/Observer.php:157]:
```suggestion
                    Mage::getResourceModel('salesrule/coupon_usage')->updateCustomerCouponTimesUsed($customerId, $coupon->getId(), true);
```

File: app/code/core/Mage/SalesRule/Model/Resource/Coupon/Usage.php [hunk 2]
```diff
@@ -46,11 +46,11 @@ protected function _construct()
     /**
      * Increment times_used counter
      *
-     *
      * @param unknown_type $customerId
      * @param unknown_type $couponId
+     * @param bool $decrement   Decrement instead of increment times_used
      */
-    public function updateCustomerCouponTimesUsed($customerId, $couponId)
+    public function updateCustomerCouponTimesUsed($customerId, $couponId, $decrement = FALSE)
```
Comment (colinmollenhour) [inline app/code/core/Mage/SalesRule/Model/Resource/Coupon/Usage.php:53]:
```suggestion
    public function updateCustomerCouponTimesUsed($customerId, $couponId, $decrement = false)
```

File: app/code/core/Mage/SalesRule/Model/Resource/Coupon/Usage.php [hunk 3]
```diff
@@ -60,11 +60,11 @@ public function updateCustomerCouponTimesUsed($customerId, $couponId)
 
         $timesUsed = $read->fetchOne($select, array(':coupon_id' => $couponId, ':customer_id' => $customerId));
 
-        if ($timesUsed > 0) {
+        if ($timesUsed !== FALSE && $timesUsed > 0) {
```
Comment (colinmollenhour) [inline app/code/core/Mage/SalesRule/Model/Resource/Coupon/Usage.php:63]:
```suggestion
        if ($timesUsed !== false && $timesUsed > 0) {
```

**Comment by henrykbrzoska** [comment]
Hi, this logic is not obvious. Please reanalyse if used coupon should be reused.
IMO - shouldn't. Regardless of whether the order has been placed.

**Comment by sreichel** [comment]
> Hi, this logic is not obvious. Please reanalyse if used coupon should be reused.
> IMO - shouldn't. Regardless of whether the order has been placed.

Can you please explain your concerns?

If I use a voucher but cancel the order, the voucher should still be available to me.

In what cases should the voucher no longer be valid?

**Comment by henrykbrzoska** [comment]
> If I use a voucher but cancel the order, the voucher should still be available to me.

are you sure?

**Comment by sreichel** [comment]
> are you sure?
> I think only about ecommerce logic. If i give someone 'ONE time' coupon it should be 'ONE time' usable.

Yeah, but maybe I'm missing something...

For example: A customer orders an article not in stock with a one-time-voucher. Since it takes too long, he cancels the order.

Should the voucher expire?

**Comment by henrykbrzoska** [comment]
IMO yes. he should get another one if seller permit.

**Comment by sreichel** [comment]
Well ... IMO it should not :)

Just did a quick search ...

- mediamarkt/saturn voucher expires
- adidas does not expire, even after full refund

I think its easy cover both needs ... add a config section and a check to observer. Time for a PR?

**Comment by henrykbrzoska** [comment]
ok, approve. :)

I wanted to be careful

**Pull request merged**